metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_shares.py

```diff
@@ -0,0 +1,229 @@
+from datahub.ingestion.source.snowflake.snowflake_schema import SnowflakeDatabase
+from datahub.ingestion.source.snowflake.snowflake_utils import SnowflakeCommonMixin
+from datahub.metadata.com.linkedin.pegasus2avro.common import Siblings
+from datahub.metadata.com.linkedin.pegasus2avro.dataset import (
+    DatasetLineageType,
+    Upstream,
+    UpstreamLineage,
+)
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+@dataclass
+class SharedDatabase:
+    """
+    Represents shared database from current platform instance
+    This is either created from an inbound share or included in an outbound share.
+    """
+
+    name: str
+    created_from_share: bool
+
+    # This will have exactly entry if created_from_share = True
+    shares: List[str]
+
+
+class SnowflakeSharesHandler(SnowflakeCommonMixin):
+    def __init__(
+        self,
+        config: SnowflakeV2Config,
+        report: SnowflakeV2Report,
+        dataset_urn_builder: Callable[[str], str],
+    ) -> None:
+        self.config = config
+        self.report = report
+        self.logger = logger
+        self.dataset_urn_builder = dataset_urn_builder
+
+    def _get_shared_databases(
+        self, shares: Dict[str, SnowflakeShareConfig], platform_instance: Optional[str]
+    ) -> Dict[str, SharedDatabase]:
+        # this is ensured in config validators
+        assert platform_instance is not None
+
+        shared_databases: Dict[str, SharedDatabase] = {}
+
+        for share_name, share_details in shares.items():
+            if share_details.platform_instance == platform_instance:
+                if share_details.database not in shared_databases:
+                    shared_databases[share_details.database] = SharedDatabase(
+                        name=share_details.database,
+                        created_from_share=False,
+                        shares=[share_name],
+                    )
+
+                else:
+                    shared_databases[share_details.database].shares.append(share_name)
+
+            else:
+                for consumer in share_details.consumers:
+                    if consumer.platform_instance == platform_instance:
+                        shared_databases[consumer.database] = SharedDatabase(
+                            name=share_details.database,
+                            created_from_share=True,
+                            shares=[share_name],
+                        )
+                        break
+                else:
+                    self.report_warning(
+                        f"Skipping Share, as it does not include current platform instance {platform_instance}",
+                        share_name,
+                    )
+
+        return shared_databases
+
+    def get_shares_workunits(
+        self, databases: List[SnowflakeDatabase]
+    ) -> Iterable[MetadataWorkUnit]:
+        shared_databases = self._get_shared_databases(
+            self.config.shares or {}, self.config.platform_instance
```

Comment:
self.config is also used in other places in SnowflakeCommonMixin - primarily in deciding whether to lowercase urn. hence keeping self.config and refractoring a bit to avoid asserts.